### PR TITLE
tidb: Split bootstrap into two steps.

### PR DESCRIPTION
**Pull request by shenli**
Fix https://github.com/pingcap/tidb/issues/515

### CODE REVIEW

File: bootstrap.go [hunk 1]
```diff
@@ -107,30 +104,33 @@ func bootstrap(s Session) {
 	} else if !errors2.ErrorEqual(err, errors.ErrDatabaseNotExist) {
 		log.Fatal(err)
 	}
-	mustExecute(s, fmt.Sprintf("CREATE DATABASE %s;", mysql.SystemDB))
-	initUserTable(s)
-	initPrivTables(s)
-	initGlobalVariables(s)
+	doDDLWorks(s)
```
Comment (ngaut) [inline bootstrap.go]: Please add tests to handle failure between doDMLWorks and doDDLWorks.

**Comment by ngaut** [comment]
@coocood @siddontang

File: bootstrap.go [hunk 2]
```diff
@@ -92,45 +93,94 @@ const (
...
+		if strings.Contains(str, "does not exist") {
+			return false, nil
+		}
+		return false, errors.Trace(err)
+	}
+	if len(rs) != 1 {
+		return false, errors.New("Wrong number of Recordset")
+	}
+	r := rs[0]
+	row, err := r.Next()
```
Comment (coocood) [inline bootstrap.go]: row may be nil here. we should check error first before use it.

File: bootstrap.go [hunk 3]
```diff
@@ -22,10 +22,11 @@ import (
 	"runtime/debug"
 	"strings"
 
+	"github.com/juju/errors"
 	"github.com/ngaut/log"
 	"github.com/pingcap/tidb/mysql"
 	"github.com/pingcap/tidb/sessionctx/variable"
-	"github.com/pingcap/tidb/util/errors"
+	terrors "github.com/pingcap/tidb/util/errors"
```
Comment (siddontang) [inline bootstrap.go]: we use qerror like this in other places.

Comment (shenli) [inline bootstrap.go]: This will be removed after Zhou Sir's pr merged into master.
I have a TODO in this file.

File: bootstrap.go [hunk 4]
```diff
@@ -92,45 +93,94 @@ const (
...
+		return
+	}
+	doDDLWorks(s)
+	doDMLWorks(s)
+}
 
+const (
+	bootstrapedVar      = "bootstraped"
+	bootstrapedVarTrue  = "True"
+	bootstrapedVarFalse = "False"
```
Comment (coocood) [inline bootstrap.go]: It seems like "False" is not used, can it be removed?

File: bootstrap.go [hunk 5]
```diff
@@ -92,45 +93,94 @@ const (
...
+)
+
+func alreadyBootstraped(s Session) (bool, error) {
```
Comment (siddontang) [inline bootstrap.go]: checkBootstrapped seems better.

**Comment by shenli** [comment]
PTAL

**Comment by shenli** [comment]
PTAL @siddontang

File: bootstrap.go [hunk 6]
```diff
@@ -92,45 +93,99 @@ const (
...
+			return false, nil
+		}
+		return false, errors.Trace(err)
+	}
+	if len(rs) != 1 {
+		return false, errors.New("Wrong number of Recordset")
+	}
+	r := rs[0]
+	row, err := r.Next()
+	if err != nil {
```
Comment (siddontang) [inline bootstrap.go]: if err != nil || row == nil {
    return false, errors.Trace(err)
}

**Comment by coocood** [comment]
LGTM

File: bootstrap.go (hunk 6 above)
Comment (shenli) [inline bootstrap.go]: And the following return will be  row.Data[0].(string) == bootstrappedVarTrue, errors.Trace(err)
I think check err first maybe clearer.

Comment (siddontang) [inline bootstrap.go]: no, the following will still return row.Data[0].(string) == bootstrappedVarTrue, nil, 

if err != nil || row == nil {
return false, errors.Trace(err)
}

has already handled the error first.

Comment (shenli) [inline bootstrap.go]: Yes, you are right.
PTAL

**Comment by shenli** [comment]
PTAL @siddontang

**Comment by siddontang** [comment]
LGTM

**Pull request merged**